syntax = "proto3";

// /usr/local/bin/protoc -I=. --cpp_out=. game.proto
// /usr/local/bin/protoc -I=. --grpc_out=. --plugin=protoc-gen-grpc=/home/dkwapisz/grpc/cmake/build/grpc_cpp_plugin game.proto
// python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. ./game.proto

package GameMessage;

service StateActionExchange {
  rpc StateAction (State) returns (Action) {}

  rpc StateReset (State) returns (Reset) {}
}

message State {
  message Direction {
    bool up = 1;
    bool right = 2;
    bool down = 3;
    bool left = 4;
  }

  bool closestObstacleIsBox = 1;
  bool allCoinsCollected = 2;
  bool coinInFoV = 3;
  Direction closestObstacle = 4;
  Direction closestCoin = 5;
  Direction closestEnemy = 6;
  Direction finishDirection = 7;
  int32 reward = 8;
  int32 clockTime = 9;
  bool gameOver = 10;
  int32 iteration = 11;
}

message Action {
  enum ActionDirection {
    IDLE = 0;
    UP = 1;
    RIGHT = 2;
    DOWN = 3;
    LEFT = 4;
  }

  ActionDirection moveDirection = 1;
  ActionDirection shotDirection = 2;
}

message Reset {
  bool resetNeeded = 1;
}